tempesta_fw/sched.c

```diff
@@ -25,87 +25,114 @@
 #include "sched.h"
 #include "tempesta_fw.h"
 
+/*
+ * Normally, schedulers are separate modules. Schedulers register
+ * and deregister themselves via register()/unregister() functions.
+ * Registered schedulers are placed on the list and can be used by
+ * Tempesta. The list is traversed on each HTTP message in search
+ * for an outgoing connection, so the list traversal operation is
+ * critical for speed. The event of a scheduler's registration or
+ * deregistration is rare. The list of schedulers is traversed on
+ * multiple CPUs simultaneously. These properties and the goals
+ * make RCU locks a beneficial choice.
+ */
+
 static LIST_HEAD(sched_list);
-static DEFINE_RWLOCK(sched_lock);
+static DEFINE_SPINLOCK(sched_lock);
 
+/*
+ * Find an outgoing connection for an HTTP message.
+ *
+ * Where an HTTP message goes in controlled by schedulers. It may
+ * or may not depend on properties of HTTP message itself. In any
+ * case, schedulers are polled in sequential order until a result
+ * is received. Schedulers that distribute HTTP messages among
+ * server groups come first in the list. The search stops when
+ * these schedulers run out.
+ *
+ * This function is always called in SoftIRQ context.
+ */
 TfwConnection *
 tfw_sched_get_srv_conn(TfwMsg *msg)
 {
 	TfwConnection *conn;
 	TfwScheduler *sched;
 
-	read_lock(&sched_lock);
-
-	list_for_each_entry(sched, &sched_list, list) {
+	rcu_read_lock();
+	list_for_each_entry_rcu(sched, &sched_list, list) {
 		if (!sched->sched_grp)
 			break;
-
-		/* Try all able schedulers until some of them gets a result. */
 		conn = sched->sched_grp(msg);
 		if (conn) {
-			read_unlock(&sched_lock);
+			rcu_read_unlock();
 			return conn;
 		}
 	}
-
-	read_unlock(&sched_lock);
+	rcu_read_unlock();
 
 	return NULL;
 }
 
-/**
+/*
  * Lookup a scheduler by name.
- * If the @name is NULL, then the first available scheduler is returned.
- * Useful only for configuration routines.
+ *
+ * If @name is NULL, then the first available scheduler is returned.
+ * Called only in user context, and used in configuration routines.
  */
 TfwScheduler *
 tfw_sched_lookup(const char *name)
 {
 	TfwScheduler *sched;
 
-	read_lock_bh(&sched_lock);
-
-	list_for_each_entry(sched, &sched_list, list) {
+	rcu_read_lock();
```

Comment:
Why not ```rcu_read_lock_bh()```? The function can be called in process context (configuration) and softirq.